snorkel/contrib/deps/labeling/model/label_model.py

```diff
@@ -0,0 +1,180 @@
+
+            # Calculate metrics, log, and checkpoint as necessary
+            metrics_dict = self._execute_logging(loss)
+            metrics_hist.update(metrics_dict)
+
+            # Update learning rate
+            self._update_lr_scheduler(epoch)
+
+    def fit_with_deps(
+        self,
+        L_train: np.ndarray,
+        Y_dev: Optional[np.ndarray] = None,
+        deps: Optional[List[Tuple[int, int]]] = None,
+        class_balance: Optional[List[float]] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Train label model using dependencies (if given).
+
+        Train label model to estimate mu, the parameters used to combine LFs.
+
+        Parameters
+        ----------
+        L_train
+            An [n,m] matrix with values in {-1,0,1,...,k-1}
+        Y_dev
+            Gold labels for dev set for estimating class_balance, by default None
+        deps
+            Optional list of pairs of correlated LF indices.
+        class_balance
+            Each class's percentage of the population, by default None
+        **kwargs
+            Arguments for changing train config defaults
+
+        Raises
+        ------
+        Exception
+            If loss in NaN
+
+        Examples
+        --------
+        >>> L = np.array([[0, 0, -1], [-1, 0, 1], [1, -1, 0]])
+        >>> Y_dev = [0, 1, 0]
+        >>> label_model = DependencyAwareLabelModel(verbose=False)
+        >>> label_model.fit_with_deps(L, deps=[(0, 2)])  # doctest: +SKIP
+        >>> label_model.fit_with_deps(L, deps=[(0, 2)], Y_dev=Y_dev)  # doctest: +SKIP
+        >>> label_model.fit_with_deps(L, deps=[(0, 2)], class_balance=[0.7, 0.3])  # doctest: +SKIP
+        """
+        # Set random seed
+        self.train_config: TrainConfig = merge_config(  # type:ignore
+            TrainConfig(), kwargs  # type:ignore
+        )
+        # Update base config so that it includes all parameters
+        random.seed(self.train_config.seed)
+        np.random.seed(self.train_config.seed)
+        torch.manual_seed(self.train_config.seed)
+
+        L_shift = L_train + 1  # convert to {0, 1, ..., k}
+        if L_shift.max() > self.cardinality:
+            raise ValueError(
+                f"L_train has cardinality {L_shift.max()}, cardinality={self.cardinality} passed in."
+            )
+
+        self._set_constants(L_shift)
+        self._set_class_balance(class_balance, Y_dev)
+        self._set_dependencies(deps or [])
+        lf_analysis = LFAnalysis(L_train)
+        self.coverage = lf_analysis.lf_coverages()
+
+        # Compute O and initialize params
+        if self.config.verbose:  # pragma: no cover
+            logging.info("Computing O...")
+        self._generate_O(L_shift)
+        self._init_params()
+
+        # Set model to train mode
+        self.train()
+
+        # Move model to GPU
+        if self.config.verbose and self.config.device != "cpu":  # pragma: no cover
+            logging.info("Using GPU...")
```

Comment:
It's worth checking on GPU - there were some parameters that had to be cast explicitly for the independent LabelModel

Reply:
You mean testing? How do we do that?

Reply:
I manually changed device to “gpu” in the tests and ran it on a machine with GPUs